File: ABC/abc390/e/main.py
n, x = map(int, input().split())
v = []
a = []
c = []
for _ in range(n):
    vi, ai, ci = map(int, input().split())
    v.append(vi)
    a.append(ai)
    c.append(ci)

ones = []
twos = []
threes = []
for i in range(n):
    if v[i] == 1:
        ones.append((v[i], a[i], c[i]))
    elif v[i] == 2:
        twos.append((v[i], a[i], c[i]))
    else:
        threes.append((v[i], a[i], c[i]))

from decimal import *

ones.sort(key=lambda x: Decimal(x[1]) / Decimal(x[2]), reverse=True)
twos.sort(key=lambda x: Decimal(x[1]) / Decimal(x[2]), reverse=True)
threes.sort(key=lambda x: Decimal(x[1]) / Decimal(x[2]), reverse=True)

arr = [ones, twos, threes]
idxes = [0, 0, 0]
sums = [0, 0, 0]
now_calories = 0
while now_calories <= x:
    min_idx = -1
    min_val = 10**9
    for i in range(3):
        if idxes[i] < len(arr[i]) and sums[i] < min_val:
            min_val = sums[i]
            min_idx = i
    if min_idx == -1:
        break
    if now_calories + arr[min_idx][idxes[min_idx]][2] > x:
        break
    sums[min_idx] += arr[min_idx][idxes[min_idx]][1]
    now_calories += arr[min_idx][idxes[min_idx]][2]
    idxes[min_idx] += 1

print(min(sums))

File: ABC/abc390/e/main.cpp
#include <bits/stdc++.h>
#include <atcoder/all>
using namespace std;
using namespace atcoder;

#define all(v) v.begin(), v.end()
#define SORT(v) sort(v.begin(), v.end())
#define RSORT(v) sort(v.rbegin(), v.rend())
#define REVERSE(v) reverse(v.begin(), v.end())
#define ll long
#define ld long double

#define int ll

int32_t main() {
    int n, x;
    cin >> n >> x;
    vector<int> v(n), a(n), c(n);
    for (int i = 0; i < n; i++) {
        cin >> v[i] >> a[i] >> c[i];
    }
    vector<tuple<int, int, int>> ones, twos, threes;
    for (int i = 0; i < n; i++) {
        if (v[i] == 1) {
            ones.push_back({v[i], a[i], c[i]});
        } else if (v[i] == 2) {
            twos.push_back({v[i], a[i], c[i]});
        } else {
            threes.push_back({v[i], a[i], c[i]});
        }
    }
    sort(all(ones), [](tuple<int, int, int> a, tuple<int, int, int> b) {
        return get<2>(a) < get<2>(b);
    });
    sort(all(twos), [](tuple<int, int, int> a, tuple<int, int, int> b) {
        return get<2>(a) < get<2>(b);
    });
    sort(all(threes), [](tuple<int, int, int> a, tuple<int, int, int> b) {
        return get<2>(a) < get<2>(b);
    });
}
